core/src/main/java/org/apache/iceberg/hadoop/HadoopCatalog.java

```diff
@@ -200,7 +210,101 @@ public void renameTable(TableIdentifier from, TableIdentifier to) {
+    }
+
+    Path nsPath = new Path(warehouseLocation, SLASH.join(namespace.levels()));
+    FileSystem fs = Util.getFs(nsPath, conf);
+
+    try {
+      if (isNamespace(fs, nsPath)) {
+        throw new AlreadyExistsException("Namespace '%s' already exists!", namespace);
+      }
+
+      fs.mkdirs(nsPath);
+
+    } catch (IOException e) {
+      throw new RuntimeIOException(e, "Create namespace failed: %s", namespace);
+    }
   }
 
+  @Override
+  public List<Namespace> listNamespaces(Namespace namespace) {
+    Path nsPath = new Path(SLASH.join(warehouseLocation, SLASH.join(namespace.levels())));
+    FileSystem fs = Util.getFs(nsPath, conf);
+
+    if (!isNamespace(fs, nsPath)) {
+      throw new NoSuchNamespaceException("Namespace does not exist: %s", namespace);
+    }
+
+    try {
+      return Stream.of(fs.listStatus(nsPath))
+        .map(FileStatus::getPath)
+        .filter(path -> isNamespace(fs, path))
+        .map(path -> append(namespace, path.getName()))
+        .collect(Collectors.toList());
+    } catch (IOException ioe) {
+      throw new RuntimeIOException(ioe, "Failed to list namespace under: %s", namespace);
+    }
+  }
+
+  @Override
+  public boolean dropNamespace(Namespace namespace) {
+    Path nsPath = new Path(SLASH.join(warehouseLocation, SLASH.join(namespace.levels())));
+    FileSystem fs = Util.getFs(nsPath, conf);
+
+    try {
+      if (!isNamespace(fs, nsPath) || namespace.isEmpty()) {
+        throw new NoSuchNamespaceException("Namespace does not exist: %s", namespace);
+      }
+
+      return fs.delete(nsPath, true /* recursive */);
+
+    } catch (IOException e) {
+      throw new RuntimeIOException(e, "Namespace delete failed: %s", namespace);
+    }
+  }
+
+  @Override
+  public boolean alterNamespace(Namespace namespace, NamespaceChange... changes) {
+    throw new UnsupportedOperationException(
+        "Cannot alter namespace " + namespace + " : alterNamespace is not supported");
+  }
+
+  @Override
+  public Map<String, String> loadNamespaceMetadata(Namespace namespace) {
+    Path nsPath = new Path(SLASH.join(warehouseLocation, SLASH.join(namespace.levels())));
+    FileSystem fs = Util.getFs(nsPath, conf);
+
+    if (!isNamespace(fs, nsPath) || namespace.isEmpty()) {
+      throw new NoSuchNamespaceException("Namespace does not exist: %s", namespace);
+    }
+
+    return ImmutableMap.of("location", nsPath.toString());
+  }
+
+  private boolean isNamespace(FileSystem fs, Path path) {
+    Path metadataPath = new Path(path, "metadata");
+    try {
+      return fs.isDirectory(path) && !(fs.exists(metadataPath) && fs.isDirectory(metadataPath) &&
+          (fs.listStatus(metadataPath, TABLE_FILTER).length >= 1));
+
+    } catch (IOException ioe) {
+      throw new RuntimeIOException(ioe, "Failed to list namespace info: %s ", path);
+    }
+  }
+  static Namespace append(Namespace ns, String name) {
```

Comment:
Is this used elsewhere, or can it be private?

Also, please move this static helper after instance methods.

Reply:
Ok fixed it to private